src/librustc_mir/interpret/step.rs

```diff
@@ -143,6 +143,12 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
         use rustc_middle::mir::Rvalue::*;
         match *rvalue {
+            Tls(did) => {
+                let id = M::thread_local_alloc_id(self, did)?;
+                let val = Scalar::Ptr(self.tag_global_base_pointer(id.into()));
```

Comment:
So when `tag_global_base_pointer` does canonicalization, this seems somewhat redundant?

But OTOH, Stacked Borrows should have a different tag for each instance of a thread-local. Hm.

Reply:
Maybe the hook should be `thread_local_pointer`, and return a `Pointer<M::PointerTag>`.

Reply:
I expect the `thread_local_alloc_id` hook to disappear once we make canonicalization eager. If I do your suggested change now, we either need to undo it in the future or will have code duplication because we need to run canonicalization inside `thread_local_pointer`

Reply:
I would expect some hook specialized to thread-locals to stay. The compile-time engines need an easy way to bail out when seeing a thread-local, and they do remain an inherent feature of the language that it makes sense to reflect in the structure of the interpreter.

Maybe this can be merged, maybe `tag_global_base_pointer` should take an `Option<DefId>` for pointers to statics (or maybe it should take *either* an `AllocId` *or* a `DefId`), I am not sure how this will look in the future.

It seems odd you are concerned about undoing something when saying in the same paragraph that you want to undo adding the new gook. ;)

Reply:
Right, I had some assumption there that we'd keep an `AllocId` for the initializer of thread locals, but since I just nixed that `AllocId` in 3d2a4ab we can't do that anymore, so we do need the hook as you correctly stated. But even in that case, why would we need it to return anything but an `AllocId` if we run the tagging right here anyway? We could call `tag_global_base_pointer` inside `thread_local_pointer` and then not have to run it here, you're right. I'm fine either way, It's just moving a `tag_global_base_pointer` call from here to miri.

Reply:
Hm, indeed `tag_global_base_pointer` cannot even get a TLS static as argument if they have no `AllocId` (except for the "hidden" `GlobalMemory::Memory` that the machine should never see here). So probably it cannot make anything with TLS.

What you have seems fine for now, we can always adjust later.